streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -671,6 +671,11 @@ public static StreamThread create(final InternalTopologyBuilder builder,
             originalReset = (String) consumerConfigs.get(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG);
             consumerConfigs.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "none");
         }
+
+        // Starting from 2.3, stream thread will always use thread client id to overwrite group.instance.id.
```

Comment:
I think we should remove this part from Streams first. There are some open questions that I've in mind and needs to potentially create a new Streams KIP for it. cc @mjsax

Reply:
After offline discussion with @guozhangwang, we sort out following key points:
1) Will the static membership affect unit test independence?
Short answer: no. The reason is because without explicitly setting the `client.id` config for stream instances, the static member id will be changed throughout the restarts since we add a random hash to `client.id`. It will essentially behave the same as current dynamic membership. Also one another confusion was that we are changing *max session timeout cap* to 30 min, instead of *default session timeout* which will remain as 10 s for either static or dynamic member. So the out-dated members will be kicked out in 10 seconds as expected.

2) The concern about thread id change throughout restarts. This is a valid concern in case where we configure two stream jobs within one JVM, so the threads will sometime shuffle from job A to job B, which unfortunately breaks the expectation of persistent thread-id numbers. This, however, shall not block us from enabling static membership for streams because the worst case is just doing repetitive rebalances as current dynamic membership. We could choose to address this application layer problem in another diff.

The conclusion is that, it does no harm to enable static membership on streams, we are just realizing there are more subtle cases we need to handle.

Let me know if this addresses your concern, thanks!

Reply:
@abbccdda I'd still suggest we add streams logic leveraging static members in this PR for further discussion than rush it in this PR.

For people who wants to use the feature in Streams asap, they can still do it by manually set the group.instance.id via consumer config prefix in StreamsConfig. But we need to think through all the cases before making it turned on by default in Streams.

Reply:
@guozhangwang I see. However it's currently not possible, since they need to have access to stream internal to set `group.instance.id` config.

Reply:
Got it, that makes sense. I think we would consider fixing the following the static stream-thread suffix number first, and then requiring users who wants to turn on static membership to specify the client-id then (otherwise internally created client-id would never be the same across lives of a streams instance). I saw you've created JIRA tickets for these tasks.